### PR TITLE
Nonpar

### PR DESCRIPTION
**Pull request by bvarjavand**
Nonparametric testing is important for graph statistics.
It is very flexible and works on graphs with unlabeled vertices as well as different numbers of vertices.

The current ASE only corrects for reflections across axes using a median heuristic, as otherwise it would take 2^d iterations where d is the larger ZG2 elbow of the graphs.

Currently by default we use a gaussian kernel with bandwidth 0.5 for pairwise distances, but this can be tuned easily.

### CODE REVIEW

File: graspy/inference/nonpar.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+            self.bandwidth = 0.5
+        return np.exp(-0.5 * np.sum(diffs ** 2, axis=2) / self.bandwidth ** 2)
+
+    def _statistic(self, X, Y):
+        N, _ = X.shape
+        M, _ = Y.shape
+        x_stat = np.sum(self._gaussian_covariance(X, X) - np.eye(N)) / (N * (N - 1))
+        y_stat = np.sum(self._gaussian_covariance(Y, Y) - np.eye(M)) / (M * (M - 1))
+        xy_stat = np.sum(self._gaussian_covariance(X, Y)) / (N * M)
+        return x_stat - 2 * xy_stat + x_stat
```
Comment (j1c) [inline graspy/inference/nonpar.py]: Should this be `x_stat - 2 * xy_stat + y_stat` not `x_stat - 2 * xy_stat + x_stat`?

Comment (bvarjavand) [reply graspy/inference/nonpar.py]: yes

**Comment by j1c** [comment]
Still bunch of things to do:
- [ ] Review the two comments in the code.
- [ ] Allow for LSE. Thinking about it more, I dont think omni makes sense here.
- [ ] Can you point me to a notebook that can be included in the documentation as a tutorial?

File: graspy/inference/nonpar.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        Bandwidth to use for gaussian kernel. If None,
+        the median heuristic will be used.
+    """
+
+    def __init__(
+        self, embedding="ase", n_components=None, n_bootstraps=200, bandwidth=None
+    ):
+
+        if type(n_bootstraps) is not int:
+            raise TypeError()
```
Comment (j1c) [inline graspy/inference/nonpar.py]: Please provide more meaningful error message.

File: graspy/inference/nonpar.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+        if bandwidth is not None and type(bandwidth) is not float:
+            raise TypeError()
```
Comment (j1c) [inline graspy/inference/nonpar.py]: Please provide more meaningful error message.

File: graspy/inference/nonpar.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        Returns
+        -------
+        p : float
+            The p value corresponding to the specified hypothesis test
+        """
+        A1 = symmetrize(import_graph(A1))
+        A2 = symmetrize(import_graph(A2))
+        # if type(A1) != 'numpy.ndarray' or type(A2) != 'numpy.ndarray':
+        #     return TypeError()
+        A1_d = select_dimension(A1)[0][-1]
```
Comment (j1c) [inline graspy/inference/nonpar.py]: Wrap this around if self.n_components is None. This doesn't allow users to set the embedding dimensions.

File: graspy/inference/nonpar.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        A2_d = select_dimension(A2)[0][-1]
+        max_d = max(A1_d, A2_d)
+        if self.embedding == "ase":
+            X1_hat = self._ase(A1, max_d)
+            X2_hat = self._ase(A2, max_d)
+        elif self.embedding == "lse":
```
Comment (j1c) [inline graspy/inference/nonpar.py]: Write a test for LSE. You will see that it will fail.

**Comment by j1c** [comment]
See comments. 

Additionally, please don't ever touch other config files other than gitignore.

**Pull request merged**